pkg/server/translation_test.go

```diff
@@ -0,0 +1,140 @@
+}
+
+func Test_fuzzyTranslate(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	translateSrv := NewTranslateServiceServer(nil, &mockTranslator{})
+
+	originalTranslation1 := randOriginalTranslation(3)
+	originalTranslation2 := randOriginalTranslation(10)
+
+	tests := []struct {
+		name               string
+		originalTranslation   *model.Translation
+		translations []model.Translation
+	}{
+		{
+			name:               "Fuzzy translate untranslated messages for one translation",
+			originalTranslation:   originalTranslation1,
+			translations: randTranslations(1, 3, originalTranslation1),
+		},
+		{
+			name:               "Fuzzy translate untranslated messages for five translations",
+			originalTranslation:   originalTranslation2,
+			translations: randTranslations(5, 5, originalTranslation2),
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			allTranslations := make(model.Translations, 0, len(tt.translations)+1)
+			allTranslations = append(allTranslations, *tt.originalTranslation)
+			allTranslations = append(allTranslations, tt.translations...)
+
+			untranslatedMessageIDLookup := make(map[string]struct{})
+
+			// Randomly set message status to untranslated
+			for _, msg := range tt.originalTranslation.Messages {
+				if gofakeit.Bool() {
+					untranslatedMessageIDLookup[msg.ID] = struct{}{}
+				}
+			}
+
+			for i := range allTranslations {
+				if allTranslations[i].Original {
+					continue
+				}
+
+				for j := range allTranslations[i].Messages {
+					if _, ok := untranslatedMessageIDLookup[allTranslations[i].Messages[j].ID]; ok {
+						allTranslations[i].Messages[j].Status = model.MessageStatusUntranslated
+					}
+				}
+			}
+
+			err := translateSrv.fuzzyTranslate(ctx, allTranslations)
+			require.NoError(t, err)
+
+			// Check that untranslated messages have been translated and marked as fuzzy for all translations.
+			for _, translation := range allTranslations {
+				if translation.Original {
+					require.Equal(t, *tt.originalTranslation, translation)
+					continue
+				}
+
+				for _, message := range translation.Messages {
+					if _, ok := untranslatedMessageIDLookup[message.ID]; ok {
+						require.Equal(t, mockTranslation, message.Message)
+						require.Equal(t, model.MessageStatusFuzzy.String(), message.Status.String())
+					} else {
+						require.Equal(t, model.MessageStatusTranslated.String(), message.Status.String())
+					}
+				}
+			}
+		})
+	}
+}
+
+// helpers
+
+// randOriginalTranslation creates a random translation with the original flag set to true.
+func randOriginalTranslation(messageCount uint) *model.Translation {
+	return rand.ModelTranslation(
+		messageCount,
+		[]rand.ModelMessageOption{rand.WithStatus(model.MessageStatusTranslated)},
+		rand.WithOriginal(true),
+		rand.WithLanguage(language.English))
+}
+
+// randTranslations creates a random translation with the original flag set to false
```

Comment:
// randTranslations creates random translations with the original flag set to false